### PR TITLE
enhancement/15141-klinger-oscilator

### PR DESCRIPTION
**Pull request by karolkolodziej**
Added new technical indicator, Klinger Oscillator. See #15141.

### CODE REVIEW

File: ts/Stock/Indicators/Klinger/KlingerOptions.d.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerOptions.d.ts:29]: Can we use the same name as for MACD signal line? See: https://api.highcharts.com/highstock/series.macd.signalLine

File: samples/unit-tests/indicator-klinger/recalculations/demo.js
Comment (pawelfus) [inline samples/unit-tests/indicator-klinger/recalculations/demo.js:207]: Fast Stochastic? 🤔

File: samples/stock/indicators/klinger/demo.js
Comment (pawelfus) [inline samples/stock/indicators/klinger/demo.js:21]: Oscillator*

File: ts/masters/indicators/klinger.src.ts
Comment (pawelfus) [reply ts/masters/indicators/klinger.src.ts:5]: And `ema`?

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:422]: Oscillator*

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:96]: `#plotOptions.sma.color` or `#plotOptions.klinger.color`?

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:36]: Oscillator*

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:50]: Oscillator*

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:46]: Oscillator*

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:349]: It seems `klingerPoint` variable is not necessary?

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:344]: Clever! :)

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:185]: In general, when `volumeSeries` was not found, we throw an error in all other technical indicators. Please, do the same, so developers get a clear message why the indicator was not rendered

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:180]: Is it necessary to define this function on every indicator recalc? If this is a utility method, then move it to the top scope

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:270]: I'm a bit confused here. The method is called `getSMA()` but returns the sum of points (`accumulatePeriodPoints` returns the sum, not the average).

**Review by pawelfus — CHANGES_REQUESTED** [review]
See comments below

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (raf18seb) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:201]: Just a suggestion:
```js
return correctFloat(
    DM + (trend === previousTrend ? previousCM : prevoiusDM)
);
```

Your call which one you choose

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (raf18seb) [reply ts/Stock/Indicators/Klinger/KlingerIndicator.ts:270]: It is divided by `period`. However, it exceeds 80 characters, so maybe we've finally found a case where too long lines can be confusing :D

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (raf18seb) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:232]: 80

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (raf18seb) [inline ts/Stock/Indicators/Klinger/KlingerIndicator.ts:345]: 80

**Review by raf18seb — CHANGES_REQUESTED** [review]

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [reply ts/Stock/Indicators/Klinger/KlingerIndicator.ts:270]: Oh boy.. Indeed, there's `/ period` at the end 😅

File: ts/masters/indicators/klinger.src.ts
Comment (karolkolodziej) [reply ts/masters/indicators/klinger.src.ts:2]: In every other indicator, there is `Highstock`, I think we might treat it as a separate task to change the names everywhere.

File: ts/masters/indicators/klinger.src.ts
Comment (karolkolodziej) [reply ts/masters/indicators/klinger.src.ts:5]: Yes, Klinger needs the `indicators` and `ema`, however, when I look to the other indicators which require others to work, they are not listed there (`idnicator.src.ts`). I'm not sure if this should be added here?

**Review by raf18seb — APPROVED** [review]
Brilliant job!

File: ts/masters/indicators/klinger.src.ts
Comment (pawelfus) [reply ts/masters/indicators/klinger.src.ts:5]: Hmm.. good question, any insight @bre1470 ?

File: ts/Stock/Indicators/Klinger/KlingerIndicator.ts
Comment (pawelfus) [reply ts/Stock/Indicators/Klinger/KlingerIndicator.ts:97]: Is signalLine's color inherited from SMA? I believe it should be inherited from Klinger indicator?

File: ts/Stock/Indicators/Klinger/KlingerOptions.d.ts
Comment (pawelfus) [inline ts/Stock/Indicators/Klinger/KlingerOptions.d.ts:29]: Is the type correct?

**Review by pawelfus — CHANGES_REQUESTED** [review]
Minor polishing and awaiting answer from @bre1470

File: ts/masters/indicators/klinger.src.ts
Comment (pawelfus) [reply ts/masters/indicators/klinger.src.ts]: Good idea! Will you create the issue tthen? :)

Comment (bre1470) [reply ts/masters/indicators/klinger.src.ts]: The dependent series types are imported over the series registry, which means that they will not be imported by tree shaking. If the Klinger indicator should work by its own, SMA and EMA have to be imported in the master file as well.

File: ts/masters/indicators/klinger.src.ts
Comment (bre1470) [reply ts/masters/indicators/klinger.src.ts:5]: The dependent series types are imported over the series registry, which means that they will not be imported by tree shaking. If the Klinger indicator should work by its own, SMA and EMA have to be imported in the master file as well.

Comment (karolkolodziej) [reply ts/masters/indicators/klinger.src.ts:5]: I think we should stick to the current convention because none of the indicators should work alone.

Comment (bre1470) [reply ts/masters/indicators/klinger.src.ts:5]: If we stick with the current conventions, the doclet requires are enough and the master file needs just the single import of Klinger indicator.

File: samples/unit-tests/indicator-klinger/recalculations/demo.js
Comment (bre1470) [inline samples/unit-tests/indicator-klinger/recalculations/demo.js:163]: Can we reduce this? In general unit tests should be as fast/minimal as possible.

File: ts/masters/indicators/klinger.src.ts
Comment (pawelfus) [reply ts/masters/indicators/klinger.src.ts:5]: Thanks! Could you create @karolkolodziej - could you create a ticket for this? :)

**Review by pawelfus — APPROVED** [review]
Approved 👍

**Review by bre1470 — CHANGES_REQUESTED** [review]
Can we reduce the test data to improve performance?

File: samples/unit-tests/indicator-klinger/recalculations/demo.js
Comment (karolkolodziej) [reply samples/unit-tests/indicator-klinger/recalculations/demo.js]: In the case of that indicator, we have to have about 70 points to calculate each line. I removed some points, but that's all I could do.

Comment (bre1470) [reply samples/unit-tests/indicator-klinger/recalculations/demo.js]: I see. Thanks!

**Review by bre1470 — APPROVED** [review]
Thank you! Looks good. 👍

**Review by pawellysy — APPROVED** [review]
Nothing to add!

**Review by TorsteinHonsi — APPROVED** [review]
Excellent, thanks! Now only awaiting CI tests before the final merge.

**Pull request merged**
